framework/src/physics/PhysicsBase.C

```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+        _mesh->getSubdomainNames(mesh_blocks_id_vec);
+    std::set<SubdomainName> new_blocks_set(mesh_blocks_names_vec.begin(),
+                                           mesh_blocks_names_vec.end());
+    blocks_set = new_blocks_set;
+  }
+
+  // Check whether there are any ICs for this variable already in the problem
+  if (isVariableFV(var_name))
+    has_all_blocks = _problem->getFVInitialConditionWarehouse().hasObjectsForVariableAndBlocks(
+        var_name, blocks_set, blocks_covered, 0);
+  else
+    has_all_blocks = _problem->getInitialConditionWarehouse().hasObjectsForVariableAndBlocks(
+        var_name, blocks_set, blocks_covered, 0);
+
+  const bool has_some_blocks = !blocks_covered.empty();
+  if (!has_some_blocks)
+    return true;
+
+  if (has_all_blocks)
+  {
+    if (error_if_already_defined)
+      mooseError("ICs for variable '" + var_name + "' have already been defined for blocks '" +
+                 Moose::stringify(blocks) + "'.");
+    else
+      return false;
+  }
+
+  // Not implemented.
+  mooseError("There is a partial overlap between the subdomains covered by pre-existing initial "
+             "conditions (ICs), defined on blocks: " +
+             Moose::stringify(blocks_covered) + "\n and a newly created IC for variable " +
+             var_name + ", to be defined on blocks: " + Moose::stringify(blocks) +
+             ".\n We should be creating the Physics' IC only for non-covered blocks. This is not "
+             "implemented at this time.");
+}
+
+bool
+PhysicsBase::shouldCreateTimeDerivative(const VariableName & var_name,
+                                        const std::vector<SubdomainName> & blocks,
+                                        const bool error_if_already_defined) const
+{
+  // Follow the transient setting of the Physics
+  if (!isTransient())
+    return false;
+
+  // Variable is either nonlinear (FV/FE), nodal nonlinear (field of ODEs), linear, or scalar.
+  // The warehouses hosting the time kernels are different for each of these types
+  // Different type of time derivatives, not block restrictable
+  mooseAssert(!isVariableScalar(var_name),
+              "shouldCreateTimeDerivative not implemented for scalar variables");
+  mooseAssert(!_problem->hasAuxiliaryVariable(var_name),
+              "Should not be called with auxiliary variables");
+
+  // Get solver system type
+  const auto var = &_problem->getVariable(0, var_name);
+  const auto var_id = var->number();
+  const auto sys_num = var->sys().number();
+  const auto time_vector_tag =
+      (sys_num < _problem->numNonlinearSystems())
+          ? var->sys().timeVectorTag()
+          // this is not quite correct. Many kernels can contribute to RHS time vector on paper
+          : dynamic_cast<LinearSystem *>(&var->sys())->rightHandSideTimeVectorTag();
+
+  // We just use the warehouse, it should cover every time derivative object type
+  bool all_blocks_covered = true, some_block_covered = false;
```

Comment:
I've looked over three different methods with three different ways of doing this block intersection calculation. It sure would be nice if we could make this more uniform. I'm not saying it can be done for sure

Reply:
uniformized now. There are two ways roughly now. Variables are a little simpler than ICs and time derivative block restriction wise